packages/fx-core/src/plugins/resource/cicd/index.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import {
+  err,
+  UserError,
+  SystemError,
+  v2,
+  TokenProvider,
+  FxError,
+  Result,
+  Inputs,
+  Json,
+  Func,
+  ok,
+  QTreeNode,
+  ProjectSettings,
+  SingleSelectQuestion,
+  Platform,
+} from "@microsoft/teamsfx-api";
+
+import { FxResult, FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDImpl } from "./plugin";
+import { ErrorType, InternalError, NoProjectOpenedError, PluginError } from "./errors";
+import { LifecycleFuncNames } from "./constants";
+import { Service } from "typedi";
+import { ResourcePluginsV2 } from "../../solution/fx-solution/ResourcePluginContainer";
+import { ResourcePlugin, Context, DeepReadonly } from "@microsoft/teamsfx-api/build/v2";
+import {
+  githubOption,
+  azdoOption,
+  jenkinsOption,
+  ciOption,
+  cdOption,
+  provisionOption,
+  publishOption,
+  questionNames,
+} from "./questions";
+import { Logger } from "./logger";
+import { environmentManager } from "../../../core/environment";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+@Service(ResourcePluginsV2.CICDPlugin)
+export class CICDPluginV2 implements ResourcePlugin {
+  name = "fx-resource-cicd";
+  displayName = "CICD";
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  activate(projectSettings: ProjectSettings): boolean {
+    return true;
+  }
+
+  public cicdImpl: CICDImpl = new CICDImpl();
+
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    Logger.setLogger(context.logProvider);
+    return await this.cicdImpl.addCICDWorkflows(context, inputs, envInfo);
+  }
+
+  public async getQuestionsForUserTask(
+    ctx: Context,
+    inputs: Inputs,
+    func: Func,
+    envInfo: DeepReadonly<v2.EnvInfoV2>,
+    tokenProvider: TokenProvider
+  ): Promise<Result<QTreeNode | undefined, FxError>> {
+    const cicdWorkflowQuestions = new QTreeNode({
+      type: "group",
+    });
+
+    const whichProvider = new QTreeNode({
+      name: questionNames.Provider,
+      type: "singleSelect",
+      staticOptions: [githubOption, azdoOption, jenkinsOption],
+      title: "Select a CI/CD Provider",
+      default: githubOption.id,
+    });
+
+    const whichTemplate = new QTreeNode({
+      name: questionNames.Template,
+      type: "multiSelect",
+      staticOptions: [ciOption, cdOption, provisionOption, publishOption],
+      title: "Select template(s)",
+      default: [ciOption.id],
+    });
+
+    if (inputs.platform === Platform.VSCode) {
+      if (!inputs.projectPath) {
```

Comment:
Why we need to check the project path again when asking user question?

Reply:
It's kind of for robustness consideration although the possibility to throw this kind of error is low since it may be guaranteed by inner contract.